File: src/windows/AmbientContextMcp.Desktop/Resources/Strings.cs
using System.Globalization;

namespace AmbientContextMcp.Resources;

/// <summary>
/// UI 表示文字列を culture 別に持つ静的リソース。アプリ起動時の
/// <see cref="CultureInfo.CurrentUICulture"/> を 1 度だけ参照し、以降は固定。
/// 言語切替にはアプリ再起動が必要 (XAML の <c>x:Static</c> バインドは parse 時に
/// 確定するため)。.resx を使わないのは、Designer.cs 生成が VS / dotnet 双方で
/// 安定して動く保証を避け、CI / dotnet build だけで完結させるため。
/// </summary>
public static class Strings
{
    // 日本語環境のみ日本語、それ以外は全部英語にフォールバックする方針。
    // 英語環境はもちろん、ドイツ語/中国語/韓国語など第三言語の OS でも、
    // 日本語より英語の方が読める可能性が高いという前提。
    // 明示的に "ja" を選ぶには Settings → Display language で UiSettings.Language = "ja"。
    private static readonly bool IsJapanese = string.Equals(
        CultureInfo.CurrentUICulture.TwoLetterISOLanguageName,
        "ja",
        StringComparison.OrdinalIgnoreCase);

    private static string T(string ja, string en) => IsJapanese ? ja : en;

    // ----- SettingsWindow.xaml -----
    public static string WindowTitle { get; } = T(
        "Ambient Context MCP 設定",
        "Ambient Context MCP — Settings");

    public static string TabMcpServer { get; } = T(
        "MCPサーバ",
        "MCP Server");

    public static string McpServerGroup { get; } = T(
        "MCP サーバ",
        "MCP Server");

    public static string LabelStatus { get; } = T(
        "状態",
        "Status");

    public static string LabelEndpoint { get; } = T(
        "Endpoint",
        "Endpoint");

    public static string LabelToken { get; } = T(
        "Token",
        "Token");

    public static string LabelPort { get; } = T(
        "ポート",
        "Port");

    public static string ButtonCopy { get; } = T(
        "コピー",
        "Copy");

    public static string PortChangeNote { get; } = T(
        "変更はアプリ再起動後に反映",
        "Changes apply after restart");

    public static string AutoStartCheckbox { get; } = T(
        "Windows ログイン時に自動起動",
        "Start automatically on Windows login");

    public static string PersistEventLogCheckbox { get; } = T(
        "イベント履歴をディスクに保存する",
        "Persist event history to disk");

    public static string PersistEventLogNote { get; } = T(
        "再起動後も保持期間内の履歴を保てます。",
        "Keeps history across restarts within the retention window.");

    public static string CopyClaudeCodeSnippet { get; } = T(
        "Claude Code 用設定スニペットをコピー",
        "Copy Claude Code config snippet");

    public static string TabTransmission { get; } = T(
        "送信設定",
        "Transmission");

    public static string TransmissionExplanation { get; } = T(
        "MCPクライアントへ公開してよいコンテキストにチェックを入れてください。未チェックの項目は、クライアントが高い権限を要求しても送信されません。",
        "Check the contexts you allow MCP clients to read. Unchecked items are never transmitted, even if a client requests a higher scope.");

    public static string AllowAllCheckbox { get; } = T(
        "すべてのコンテキストを許可する",
        "Allow every context");

    public static string EventHistoryGroup { get; } = T(
        "イベント履歴",
        "Event history");

    public static string SensitivityLegend { get; } = T(
        "機微度: Low = 公開しても影響が小さい / Medium = 文脈次第で機微になり得る / High = 個人特定や具体的行動を含み得る",
        "Sensitivity: Low = low impact if shared / Medium = context-dependent / High = may identify the user or expose specific behavior");

    public static string LabelRetention { get; } = T(
        "保持期間",
        "Retention");

    public static string LabelMaxCount { get; } = T(
        "最大件数",
        "Max events");

    public static string Retention1Hour { get; } = T(
        "1時間",
        "1 hour");

    public static string Retention6Hours { get; } = T(
        "6時間",
        "6 hours");

    public static string Retention24Hours { get; } = T(
        "24時間",
        "24 hours");

    public static string Retention7Days { get; } = T(
        "7日",
        "7 days");

    public static string Count100 { get; } = T(
        "100件",
        "100");

    public static string Count500 { get; } = T(
        "500件",
        "500");

    public static string Count1000 { get; } = T(
        "1000件",
        "1,000");

    public static string Count5000 { get; } = T(
        "5000件",
        "5,000");

    public static string ButtonSave { get; } = T(
        "保存",
        "Save");

    public static string ButtonClose { get; } = T(
        "閉じる",
        "Close");

    // ----- Language picker -----
    public static string LabelLanguage { get; } = T(
        "表示言語",
        "Display language");

    public static string LanguageSystemDefault { get; } = T(
        "システムに従う",
        "System default");

    public static string LanguageJapanese { get; } = T(
        "日本語",
        "Japanese (日本語)");

    public static string LanguageEnglish { get; } = T(
        "English",
        "English");

    // ----- Status messages (SettingsWindow.xaml.cs) -----
    public static string StatusSaved { get; } = T(
        "保存しました。送信設定は次回の文脈更新から反映されます。ポート変更はアプリ再起動後に有効になります。",
        "Saved. Transmission settings take effect at the next context refresh. Port changes apply after restart.");

    public static string StatusSavedNeedsRestart { get; } = T(
        "保存しました。言語とポートの変更はアプリ再起動後に反映されます。",
        "Saved. Language and port changes take effect after restart.");

    public static string StatusClaudeCodeCopied { get; } = T(
        "Claude Code 用のコマンドをクリップボードにコピーしました。",
        "Copied the Claude Code command to the clipboard.");

    /// <summary>Format: {0} = exception type, {1} = message.</summary>
    public static string StatusSaveFailedFormat { get; } = T(
        "保存に失敗しました: {0}: {1}",
        "Save failed: {0}: {1}");

    /// <summary>Format: {0} = port number.</summary>
    public static string StatusMcpRunningFormat { get; } = T(
        "起動中 :{0}",
        "Running :{0}");

    // ----- Transmission option groups (SettingsWindow transmission tab) -----
    public static string TxGroupForegroundApp { get; } = T(
        "フォアグラウンドアプリ",
        "Foreground app");

    public static string TxGroupActivity { get; } = T(
        "作業リズム",
        "Work rhythm");

    public static string TxGroupMedia { get; } = T(
        "メディア",
        "Media");

    public static string TxGroupEnvironment { get; } = T(
        "環境",
        "Environment");

    public static string TxUiForegroundIdentity { get; } = T(
        "作業カテゴリ・名前・プロセス名（現在値 + 切替通知）",
        "Work category, name, and process (current value + switch notifications)");

    public static string TxUiForegroundTitleSummary { get; } = T(
        "タイトル要約（現在値 + 変更履歴）",
        "Title summary (current value + change history)");

    public static string TxUiForegroundRawTitle { get; } = T(
        "タイトル原文（現在値 + 変更履歴）",
        "Raw title (current value + change history)");

    public static string TxUiActivitySwitchRate { get; } = T(
        "切替頻度（現在値）",
        "Switch rate (current value)");

    public static string TxUiActivitySwitchBurst { get; } = T(
        "切替急増（通知）",
        "Switch burst (notification)");

    public static string TxUiMediaOverview { get; } = T(
        "再生の有無・状態・再生元（現在値 + 再生通知）",
        "Playback presence, status, and source (current value + playback notifications)");

    public static string TxUiMediaTitle { get; } = T(
        "タイトル（現在値 + 変更履歴）",
        "Title (current value + change history)");

    public static string TxUiMediaArtist { get; } = T(
        "アーティスト（現在値 + 変更履歴）",
        "Artist (current value + change history)");

    public static string TxUiMediaAlbum { get; } = T(
        "アルバム（現在値 + 変更履歴）",
        "Album (current value + change history)");

    public static string TxUiEnvironmentTimezone { get; } = T(
        "タイムゾーン（現在値 + 変更通知）",
        "Time zone (current value + change notifications)");

    public static string TxUiEnvironmentDisplays { get; } = T(
        "ディスプレイ構成（現在値 + 変更通知）",
        "Display layout (current value + change notifications)");

    // ----- Transmission options (legacy per-path labels; kept for reference) -----
    public static string TxOptForegroundCategory { get; } = T(
        "フォアグラウンドアプリの作業カテゴリ",
        "Foreground app work category");

    public static string TxOptForegroundAppName { get; } = T(
        "フォアグラウンドアプリ名",
        "Foreground app name");

    public static string TxOptForegroundProcessName { get; } = T(
        "フォアグラウンドアプリのプロセス名",
        "Foreground app process name");

    public static string TxOptForegroundTitleSummary { get; } = T(
        "フォアグラウンドウィンドウのタイトル要約",
        "Foreground window title summary");

    public static string TxOptForegroundRawWindowTitle { get; } = T(
        "フォアグラウンドウィンドウのタイトル原文",
        "Foreground window raw title");

    public static string TxOptEventForegroundChanged { get; } = T(
        "フォアグラウンドアプリ切替イベント (アプリ名・プロセス名込み)",
        "Foreground app switch event (includes app and process name)");

    public static string TxOptEventForegroundTitleChanged { get; } = T(
        "フォアグラウンドウィンドウのタイトル変更イベント (アプリ文脈込み)",
        "Foreground window title change event (includes app context)");

    public static string TxOptEventForegroundTitleChangedSummary { get; } = T(
        "フォアグラウンドウィンドウのタイトル変更イベント: 要約",
        "Foreground window title change event: summary");

    public static string TxOptEventForegroundTitleChangedRaw { get; } = T(
        "フォアグラウンドウィンドウのタイトル変更イベント: 原文",
        "Foreground window title change event: raw title");

    public static string TxOptActivityContextSwitches { get; } = T(
        "フォアグラウンドアプリ切替頻度",
        "Foreground app switch rate");

    public static string TxOptEventContextSwitchBurst { get; } = T(
        "フォアグラウンドアプリ切替増加イベント",
        "Foreground app switch burst event");

    public static string TxOptMediaIsAvailable { get; } = T(
        "メディアセッション有無",
        "Media session presence");

    public static string TxOptMediaPlaybackStatus { get; } = T(
        "メディア再生状態",
        "Media playback status");

    public static string TxOptMediaSourceApp { get; } = T(
        "メディア再生元アプリ",
        "Media source app");

    public static string TxOptMediaTitle { get; } = T(
        "メディアタイトル",
        "Media title");

    public static string TxOptMediaArtist { get; } = T(
        "メディアアーティスト",
        "Media artist");

    public static string TxOptMediaAlbumTitle { get; } = T(
        "メディアアルバム",
        "Media album");

    public static string TxOptEventMediaPlaybackStarted { get; } = T(
        "メディア再生開始イベント",
        "Media playback started event");

    public static string TxOptEventMediaPlaybackPaused { get; } = T(
        "メディア一時停止イベント",
        "Media playback paused event");

    public static string TxOptEventMediaSessionChanged { get; } = T(
        "メディアセッション変更イベント",
        "Media session changed event");

    public static string TxOptEventMediaSessionChangedTitle { get; } = T(
        "メディアセッション変更イベント: タイトル",
        "Media session changed: title");

    public static string TxOptEventMediaSessionChangedArtist { get; } = T(
        "メディアセッション変更イベント: アーティスト",
        "Media session changed: artist");

    public static string TxOptEventMediaSessionChangedAlbumTitle { get; } = T(
        "メディアセッション変更イベント: アルバム",
        "Media session changed: album");

    public static string TxOptSystemTimeZone { get; } = T(
        "タイムゾーン",
        "Time zone");

    public static string TxOptDisplayCount { get; } = T(
        "ディスプレイ数",
        "Display count");

    public static string TxOptDisplays { get; } = T(
        "ディスプレイ構成",
        "Display layout");

    // ----- Tray (TrayHost.cs) -----
    public static string TraySettings { get; } = T(
        "設定...",
        "Settings...");

    public static string TrayCopyMcpUrl { get; } = T(
        "MCP URL をコピー",
        "Copy MCP URL");

    public static string TrayCopyMcpToken { get; } = T(
        "MCP トークンをコピー",
        "Copy MCP token");

    public static string TrayCopyClaudeCodeSnippet { get; } = T(
        "Claude Code 用設定をコピー",
        "Copy Claude Code config");

    public static string TrayPause { get; } = T(
        "一時停止",
        "Pause");

    public static string TrayResume { get; } = T(
        "再開",
        "Resume");

    public static string TrayPausedSuffix { get; } = T(
        " (一時停止中)",
        " (paused)");

    public static string TrayExit { get; } = T(
        "終了",
        "Exit");

    // ----- Program.cs startup error -----
    /// <summary>Format: {0} = exception type/message, {1} = port number.</summary>
    public static string StartupErrorFormat { get; } = T(
        "Ambient Context MCP の起動に失敗しました。\n\n{0}\n\nポート {1} が他プロセスで使用中の可能性があります。\n%LOCALAPPDATA%\\AmbientContextMcp\\settings.json の mcpServer.port を変更するか、同ファイルを削除して再起動してください。",
        "Ambient Context MCP failed to start.\n\n{0}\n\nPort {1} may be in use by another process. Edit mcpServer.port in %LOCALAPPDATA%\\AmbientContextMcp\\settings.json, or delete the file and restart.");
}

File: src/macos/Sources/AmbientContextMac/App/Strings.swift
import Foundation

/// C# `AmbientContextMcp.Resources.Strings` (src/windows/AmbientContextMcp.Desktop/Resources/Strings.cs)
/// の移植。キーは C# のプロパティ名をそのまま使い、値は `Resources/{en,ja}.lproj/Localizable.strings`
/// に置く。
///
/// **`Bundle.module` は使わない。** SwiftPM が生成する `Bundle.module` は
/// `.app` 直下 → ビルド時の絶対パス (`.build/...`) の順にしか探さず、
/// 開発機でだけ通る罠になる (PoC 4 §5 / 設計書 §4 Phase 4)。ここでは
///   1. `Bundle.main.resourceURL` (= `.app/Contents/Resources`, build-app.sh が .lproj を配置)
///   2. `#filePath` から辿るソースツリー (`swift run` / テスト実行時)
/// の順に `.lproj/Localizable.strings` を直接読む。
///
/// 言語決定: 明示設定 (`ui.language` = "ja" / "en") が最優先。空なら
/// `Locale.preferredLanguages` の先頭が ja 始まりなら ja、それ以外は英語
/// (C# の「日本語環境のみ日本語、それ以外は英語」と同じ方針)。
enum Strings {
    /// C# は `CultureInfo.CurrentUICulture` を 1 度だけ見る。こちらは起動時に
    /// `configure(languageSetting:)` を 1 度呼び、以降は固定。
    private static let state = StringsState()

    /// 実際に採用した言語 ("ja" / "en")。Core のカタログ言語にも同じ値を渡す。
    static var language: String { state.language }

    /// 解決したリソースディレクトリ (診断ログ用)。
    static var resourcePath: String { state.resourcePath }

    /// - Parameter languageSetting: `UiSettings.language` の生値 ("" / "ja" / "en")。
    static func configure(languageSetting: String) {
        state.configure(languageSetting: languageSetting)
    }

    /// キー未定義なら en にフォールバックし、それも無ければキー自身を返す。
    static func text(_ key: String) -> String {
        state.text(key)
    }

    /// `%1$@` 形式のプレースホルダを埋める。C# の `string.Format` (`{0}`) に対応。
    static func format(_ key: String, _ arguments: CVarArg...) -> String {
        String(format: text(key), arguments: arguments)
    }

    // MARK: - C# Strings.cs と 1:1 のアクセサ

    static var windowTitle: String { text("WindowTitle") }
    static var tabMcpServer: String { text("TabMcpServer") }
    static var mcpServerGroup: String { text("McpServerGroup") }
    static var labelStatus: String { text("LabelStatus") }
    static var labelEndpoint: String { text("LabelEndpoint") }
    static var labelToken: String { text("LabelToken") }
    static var labelPort: String { text("LabelPort") }
    static var buttonCopy: String { text("ButtonCopy") }
    static var portChangeNote: String { text("PortChangeNote") }
    static var autoStartCheckbox: String { text("AutoStartCheckbox") }
    static var persistEventLogCheckbox: String { text("PersistEventLogCheckbox") }
    static var persistEventLogNote: String { text("PersistEventLogNote") }
    static var copyClaudeCodeSnippet: String { text("CopyClaudeCodeSnippet") }
    static var tabTransmission: String { text("TabTransmission") }
    static var transmissionExplanation: String { text("TransmissionExplanation") }
    static var allowAllCheckbox: String { text("AllowAllCheckbox") }
    static var eventHistoryGroup: String { text("EventHistoryGroup") }
    static var sensitivityLegend: String { text("SensitivityLegend") }
    static var labelRetention: String { text("LabelRetention") }
    static var labelMaxCount: String { text("LabelMaxCount") }
    static var retention1Hour: String { text("Retention1Hour") }
    static var retention6Hours: String { text("Retention6Hours") }
    static var retention24Hours: String { text("Retention24Hours") }
    static var retention7Days: String { text("Retention7Days") }
    static var count100: String { text("Count100") }
    static var count500: String { text("Count500") }
    static var count1000: String { text("Count1000") }
    static var count5000: String { text("Count5000") }
    static var buttonSave: String { text("ButtonSave") }
    static var buttonClose: String { text("ButtonClose") }

    static var labelLanguage: String { text("LabelLanguage") }
    static var languageSystemDefault: String { text("LanguageSystemDefault") }
    static var languageJapanese: String { text("LanguageJapanese") }
    static var languageEnglish: String { text("LanguageEnglish") }

    static var statusSaved: String { text("StatusSaved") }
    static var statusSavedNeedsRestart: String { text("StatusSavedNeedsRestart") }
    static var statusClaudeCodeCopied: String { text("StatusClaudeCodeCopied") }
    static func statusSaveFailed(_ type: String, _ message: String) -> String {
        format("StatusSaveFailedFormat", type, message)
    }
    /// 保存自体は成功したが、ログイン項目の適用だけ失敗したときのメインステータス行。
    /// (C# は autostart の失敗が例外になり保存全体が失敗表示になる。macOS 版は他の
    /// セクションの保存を活かしたうえで、失敗をここに出す)
    static func statusSavedWithAutostartFailure(_ saved: String, _ reason: String) -> String {
        format("MacStatusSavedAutostartFailedFormat", saved, reason)
    }
    static func statusMcpRunning(port: Int) -> String {
        format("StatusMcpRunningFormat", String(port))
    }

    static var txGroupForegroundApp: String { text("TxGroupForegroundApp") }
    static var txGroupActivity: String { text("TxGroupActivity") }
    static var txGroupMedia: String { text("TxGroupMedia") }
    static var txGroupEnvironment: String { text("TxGroupEnvironment") }

    static var traySettings: String { text("TraySettings") }
    static var trayCopyMcpUrl: String { text("TrayCopyMcpUrl") }
    static var trayCopyMcpToken: String { text("TrayCopyMcpToken") }
    static var trayCopyClaudeCodeSnippet: String { text("TrayCopyClaudeCodeSnippet") }
    static var trayPause: String { text("TrayPause") }
    static var trayResume: String { text("TrayResume") }
    static var trayPausedSuffix: String { text("TrayPausedSuffix") }
    static var trayExit: String { text("TrayExit") }

    static func startupError(_ detail: String, port: Int) -> String {
        format("StartupErrorFormat", detail, String(port))
    }

    // MARK: - カタログ由来 ID → 表示名 (C# TransmissionGroupViewModel と同じ写像)

    static func transmissionGroupTitle(groupId: String) -> String {
        switch groupId {
        case "foregroundApp": return txGroupForegroundApp
        case "activity": return txGroupActivity
        case "media": return txGroupMedia
        case "environment": return txGroupEnvironment
        default: return groupId
        }
    }

    static func transmissionOptionLabel(optionId: String) -> String {
        switch optionId {
        case "foreground.identity": return text("TxUiForegroundIdentity")
        case "foreground.titleSummary": return text("TxUiForegroundTitleSummary")
        case "foreground.rawTitle": return text("TxUiForegroundRawTitle")
        case "activity.switchRate": return text("TxUiActivitySwitchRate")
        case "activity.switchBurst": return text("TxUiActivitySwitchBurst")
        case "media.overview": return text("TxUiMediaOverview")
        case "media.title": return text("TxUiMediaTitle")
        case "media.artist": return text("TxUiMediaArtist")
        case "media.album": return text("TxUiMediaAlbum")
        case "environment.timezone": return text("TxUiEnvironmentTimezone")
        case "environment.displays": return text("TxUiEnvironmentDisplays")
        default: return optionId
        }
    }
}

/// `Strings` の可変状態。起動直後に 1 度だけ書き、以降は読むだけだが、
/// actor をまたいで読まれ得るのでロックで守る。
private final class StringsState: @unchecked Sendable {
    private let lock = NSLock()
    private var table: [String: String] = [:]
    private var fallback: [String: String] = [:]
    private var languageValue = "en"
    private var resourcePathValue = ""

    var language: String {
        lock.lock()
        defer { lock.unlock() }
        return languageValue
    }

    var resourcePath: String {
        lock.lock()
        defer { lock.unlock() }
        return resourcePathValue
    }

    func configure(languageSetting: String) {
        let resolved = StringsState.resolveLanguage(setting: languageSetting)
        let directory = StringsState.resourceDirectory()
        let loaded = directory.flatMap { StringsState.loadTable(directory: $0, language: resolved) } ?? [:]
        let english = resolved == "en"
            ? loaded
            : (directory.flatMap { StringsState.loadTable(directory: $0, language: "en") } ?? [:])

        lock.lock()
        languageValue = resolved
        resourcePathValue = directory?.path ?? ""
        table = loaded
        fallback = english
        lock.unlock()
    }

    func text(_ key: String) -> String {
        lock.lock()
        defer { lock.unlock() }
        return table[key] ?? fallback[key] ?? key
    }

    /// 明示設定が最優先。空なら OS の優先言語 (ja 始まりのみ ja)。
    static func resolveLanguage(setting: String) -> String {
        let normalized = setting.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if normalized.hasPrefix("ja") { return "ja" }
        if normalized.hasPrefix("en") { return "en" }
        let preferred = (Locale.preferredLanguages.first ?? "en").lowercased()
        return preferred.hasPrefix("ja") ? "ja" : "en"
    }

    /// `.lproj` を含むディレクトリ。バンドル実行なら Contents/Resources、
    /// `swift run` ならソースツリーの Sources/AmbientContextMac/Resources。
    static func resourceDirectory() -> URL? {
        if let resources = Bundle.main.resourceURL,
           FileManager.default.fileExists(atPath: resources.appendingPathComponent("en.lproj").path) {
            return resources
        }
        let sourceResources = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()   // App/
            .deletingLastPathComponent()   // AmbientContextMac/
            .appendingPathComponent("Resources")
        if FileManager.default.fileExists(atPath: sourceResources.appendingPathComponent("en.lproj").path) {
            return sourceResources
        }
        return nil
    }

    /// `Localizable.strings` (旧形式 plist) を辞書として読む。
    /// `NSLocalizedString` を使わないのは、バンドルの言語交渉ではなく
    /// 「設定で選んだ言語」を確実に採用するため。
    static func loadTable(directory: URL, language: String) -> [String: String]? {
        let url = directory
            .appendingPathComponent("\(language).lproj")
            .appendingPathComponent("Localizable.strings")
        guard let data = try? Data(contentsOf: url),
              let object = try? PropertyListSerialization.propertyList(from: data, format: nil),
              let dictionary = object as? [String: String] else {
            return nil
        }
        return dictionary
    }
}
